### PR TITLE
312

### PR DESCRIPTION
**Pull request by andyfaff**

### CODE REVIEW

**Comment by lesteve** [comment]
It looks like in e9238b9f0700f0dda8a1bf9cf3d70a9965b580b0 ([build log](https://github.com/andyfaff/scipy/actions/runs/5620900044/job/15230683251?pr=53)) it is still downloading numpy from source and trying to build. I can see `Downloading numpy-1.25.1.tar.gz` in the log.

Maybe `--no-build-isolation` should be used for the `python -m pip wheel` call?

**Comment by andyfaff** [comment]
@rgommers, can you parse what's going on in https://github.com/andyfaff/scipy/actions/runs/5620900044/job/15230683251? I've installed numpy with `CIBW_BEFORE_BUILD`, but the pip invocation that cibuildwheel uses seems to override the numpy install.

**Comment by andyfaff** [comment]
I think this is the command being used by pip:

> /opt/python/cp312-cp312/bin/python /opt/_internal/cpython-3.12.0b4/lib/python3.12/site-packages/pip/__pip-runner__.py install --ignore-installed --no-user --prefix /tmp/pip-build-env-pf_aph9y/overlay --no-warn-script-location -v --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'meson-python>=0.12.1' 'Cython>=0.29.35' 'pybind11>=2.10.4' 'pythran>=0.12.0' wheel 'numpy==1.22.4; python_version<='"'"'3.10'"'"' and platform_python_implementation != '"'"'PyPy'"'"'' 'numpy==1.23.2; python_version=='"'"'3.11'"'"' and platform_python_implementation != '"'"'PyPy'"'"'' 'numpy; python_version>='"'"'3.12'"'"'' 'numpy; python_version>='"'"'3.9'"'"' and platform_python_implementation=='"'"'PyPy'"'"''

**Comment by lesteve** [comment]
Maybe `CIBW_ENVIRONMENT: "PIP_NO_BUILD_ISOLATION=1"` as mentioned for example in https://github.com/pypa/cibuildwheel/issues/486#issuecomment-748469214

**Comment by rgommers** [comment]
> Maybe `CIBW_ENVIRONMENT: "PIP_NO_BUILD_ISOLATION=1"` as mentioned for example in [pypa/cibuildwheel#486 (comment)](https://github.com/pypa/cibuildwheel/issues/486#issuecomment-748469214)

That sounds right, but the last commit/log shows it's not enough yet. Maybe due to this follow-up comment: https://github.com/pypa/cibuildwheel/issues/486#issuecomment-748486020?

**Comment by rgommers** [comment]
And if build isolation is successfully disabled, then the other build dependencies must also be installed manually (I believe `wheel` is no longer needed though, so that can be left out).

**Comment by lesteve** [comment]
Maybe an alternative approach would be to set pip environment variable to that the `python -m pip wheel` command install its build dependencies (hence numpy) from scientific-python-nightly-wheels?

Something like this maybe?
```
PIP_EXTRA_INDEX_URL=https://pypi.anaconda.org/scientific-python-nightly-wheels/simple
PIP_PRE=1
```

**Comment by lesteve** [comment]
It seems like combining my previous comment with `CI_BW_ENVIRONMENT` may actually work see https://github.com/lesteve/scipy/pull/1.

**Comment by rgommers** [comment]
Awesome!

**Comment by andyfaff** [comment]
@lesteve, the suggestions seem to work for musllinux and manylinux on GHA. However, the modification doesn't work for the `cp312-macosx` entry. It still tries to build from source, and isn't obeying the environment variable in the same manner as on Linux. I checked and the environment variable is present.

**Comment by andyfaff** [comment]
I think the issues are that the environment variables wre being clobbered later on. Trying to explore that.

File: .github/workflows/wheels.yml [hunk 1]
```diff
@@ -124,11 +123,22 @@ jobs:
...
           CIBW_ENVIRONMENT_PASS_LINUX: RUNNER_OS
+          CIBW_PRERELEASE_PYTHONS: True
+
+          # required so that cp312 can grab a nightly wheel
+          # can remove when cp312 is release and we don't need to search
+          # other wheel locations.
+          CIBW_ENVIRONMENT: PIP_EXTRA_INDEX_URL=https://pypi.anaconda.org/scientific-python-nightly-wheels/simple PIP_PRE=1
+
+          CIBW_ENVIRONMENT_WINDOWS: >
+            PKG_CONFIG_PATH: c:/opt/64/lib/pkgconfig
```
Comment (andyfaff) [inline .github/workflows/wheels.yml:139]:
```suggestion
            PKG_CONFIG_PATH=c:/opt/64/lib/pkgconfig
```

**Pull request closed without merging**